src/Authenticator/PasetoAuthenticator.php

```diff
@@ -0,0 +1,196 @@
+    public function __construct(IdentifierInterface $identifier, array $config = [])
+    {
+        parent::__construct($identifier, $config);
+
+        if ($this->getConfig('version') !== null && !$this->getConfig('version') instanceof ProtocolInterface) {
+            throw new RuntimeException(
+                'PASETO `version` config must be an instance of ' . ProtocolInterface::class
+            );
+        }
+
+        if (!in_array($this->getConfig('purpose'), [self::PUBLIC, self::LOCAL])) {
+            throw new RuntimeException('PASETO `purpose` config must one of: local or public');
+        }
+
+        if (empty($this->getConfig('secretKey'))) {
+            if (!class_exists(\Cake\Utility\Security::class)) {
+                throw new RuntimeException('PASETO `secretKey` config must be defined');
+            }
+            $this->setConfig('secretKey', \Cake\Utility\Security::getSalt());
+        }
+    }
+
+    /**
+     * Authenticates the identity based on a PASETO token contained in a request.
+     *
+     * @link https://paseto.io/
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request that contains login information.
+     * @return \Authentication\Authenticator\ResultInterface
+     * @throws \Exception
+     */
+    public function authenticate(ServerRequestInterface $request): ResultInterface
+    {
+        try {
+            $result = $this->getPayload($request);
+        } catch (Exception $e) {
+            return new Result(
+                null,
+                Result::FAILURE_CREDENTIALS_INVALID,
+                [
+                    'message' => $e->getMessage(),
+                    'exception' => $e,
+                ]
+            );
+        }
+
+        if (!$result instanceof JsonToken) {
+            return new Result(null, Result::FAILURE_CREDENTIALS_INVALID);
+        }
+
+        if (empty($result->getSubject())) {
+            return new Result(null, Result::FAILURE_CREDENTIALS_MISSING);
+        }
+
+        if ($this->getConfig('returnPayload')) {
+            $array = array_merge(
+                $result->getClaims(),
+                ['footer' => $result->getFooterArray()]
+            );
+
+            return new Result(new ArrayObject($array), Result::SUCCESS);
+        }
+
+        $user = $this->_identifier->identify([
+            'sub' => $result->getSubject(),
+        ]);
+
+        if (empty($user)) {
+            return new Result(null, Result::FAILURE_IDENTITY_NOT_FOUND, $this->_identifier->getErrors());
+        }
+
+        return new Result($user, Result::SUCCESS);
+    }
+
+    /**
+     * Get payload data.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface|null $request Request to get authentication information from.
+     * @return object|null Payload object on success, null on failure
+     * @throws \Exception
+     */
+    public function getPayload(?ServerRequestInterface $request = null): ?object
+    {
+        if (!$request) {
```

Comment:
Not sure I understand this condition, it was borrowed from JwtAuthenticator. How is this scenario possible?

Reply:
The behavior looks like it wants to reuse the stored `payload` property. By passing `null` in, you get the last parsed result.